pkg/controller/controller.go

```diff
@@ -257,3 +260,112 @@ func initResource(clientset kubernetes.Interface) error {
 
 	return nil
 }
+
+//RunGC starts garbage collector
+func (c *Controller) RunGC(d time.Duration) {
+	timer := time.NewTimer(d)
+	defer timer.Stop()
+	for {
+		<-timer.C
+		err := c.garbageCollect()
+		if err != nil {
+			c.logger.Warningf("Failed to cleanup resources: %v", err)
+		}
+	}
+}
+
+func (c *Controller) garbageCollect() error {
+	functionList := spec.FunctionList{}
+	tprClient, err := utils.GetTPRClient()
+	if err != nil {
+		return err
+	}
+	err = tprClient.Get().Resource("functions").Do().Into(&functionList)
+	if err != nil {
+		return err
+	}
+
+	functionUIDSet := make(map[types.UID]bool)
+	for _, f := range functionList.Items {
+		functionUIDSet[f.Metadata.UID] = true
+	}
+
+	if err = collectServices(c.clientset, functionUIDSet); err != nil {
+		return err
+	}
+	if err = collectDeployment(c.clientset, functionUIDSet); err != nil {
+		return err
+	}
+	if err = collectConfigMap(c.clientset, functionUIDSet); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func collectServices(c kubernetes.Interface, functionUIDSet map[types.UID]bool) error {
+	srvs, err := c.CoreV1().Services(api.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, srv := range srvs.Items {
+		if len(srv.OwnerReferences) == 0 {
+			continue
+		}
+		if !functionUIDSet[srv.OwnerReferences[0].UID] {
+			err = c.CoreV1().Services(srv.GetNamespace()).Delete(srv.GetName(), nil)
+			if err != nil && !utils.IsKubernetesResourceNotFoundError(err) {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func collectDeployment(c kubernetes.Interface, functionUIDSet map[types.UID]bool) error {
+	ds, err := c.AppsV1beta1().Deployments(api.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, d := range ds.Items {
+		if len(d.OwnerReferences) == 0 {
+			continue
+		}
+		if !functionUIDSet[d.OwnerReferences[0].UID] {
+			err = c.AppsV1beta1().Deployments(d.GetNamespace()).Delete(d.GetName(), utils.CascadeDeleteOptions(0))
```

Comment:
woah, why are you forcing a graceperiod of zero seconds?  Zero grace period is a big hammer to use that could result in all sorts of orphaned resources...